blocks_reexecutor/blocks_reexecutor.go

```diff
@@ -111,32 +120,41 @@ func New(c *Config, blockchain *core.BlockChain, fatalErrChan chan error) *Block
 	if start > 0 && start != chainStart {
 		start--
 	}
-	// Divide work equally among available threads when BlocksPerThread is zero
-	if c.BlocksPerThread == 0 {
+	// Divide work equally among available threads when MinBlocksPerThread is zero
+	if c.MinBlocksPerThread == 0 {
 		// #nosec G115
 		work := (end - start) / uint64(c.Room)
 		if work > 0 {
-			blocksPerThread = work
+			minBlocksPerThread = work
 		}
 	}
-	return &BlocksReExecutor{
-		config:          c,
-		blockchain:      blockchain,
-		currentBlock:    end,
-		startBlock:      start,
-		blocksPerThread: blocksPerThread,
-		done:            make(chan struct{}, c.Room),
-		fatalErrChan:    fatalErrChan,
-		stateFor: func(header *types.Header) (*state.StateDB, arbitrum.StateReleaseFunc, error) {
-			state, err := blockchain.StateAt(header.Root)
-			return state, arbitrum.NoopStateRelease, err
-		},
+	trieConfig := triedb.Config{
+		Preimages: false,
+		HashDB:    hashdb.Defaults,
+	}
+	blocksReExecutor := &BlocksReExecutor{
+		config:             c,
+		db:                 state.NewDatabaseWithConfig(ethDb, &trieConfig),
+		blockchain:         blockchain,
+		currentBlock:       end,
+		startBlock:         start,
+		minBlocksPerThread: minBlocksPerThread,
+		done:               make(chan struct{}, c.Room),
+		fatalErrChan:       fatalErrChan,
 	}
+	blocksReExecutor.stateFor = func(header *types.Header) (*state.StateDB, arbitrum.StateReleaseFunc, error) {
+		sdb, err := state.NewDeterministic(header.Root, blocksReExecutor.db)
+		if err == nil {
+			_ = blocksReExecutor.db.TrieDB().Reference(header.Root, common.Hash{}) // Will be dereferenced later in advanceStateUpToBlock
```

Comment:
how about following the release function pattern?
 
* we could return `arbitrum.StateReleaseFunc` that will call `Dereference`. That can be something like `func() {triedb.Dereference(root)}`)

* we would need to move the call to release func returned from `arbitrum.FindLastAvailableState` to the thread started with LaunchThread  so as it would be called after `AdvanceStateUpToBlock` or even better - we could pass the `startStateRelease` to `advanceStateUpToBlock` and call it after we reference next state


* arbitrum.AdvanceStateUpToBlock was used only here, so we can modify it to commit statedb after each reexecuted block and check the hashes OR we can just remove `arbitrum.AdvanceStateUpToBlock`

* we can return release func also from `advanceStateUpToBlock` and have something like:
```

startState, startHeader, startStateRelease, err := arbitrum.FindLastAvailableState(ctx, s.blockchain, s.stateFor, s.blockchain.GetHeaderByNumber(start), nil, -1)
...
s.LaunchThread(func(ctx context.Context) {
    ...
    _, release, err := advanceStateUpToBlock(..., startHeader, startStateRelease, ...)
    ...
    defer release()
    ...
}
```

Reply:
note to `startStateRelease` - we should make sure that it will be called also in case of an error

Reply:
Agreed, its a great idea to utilize `StateReleaseFunc`s here! 
Also removed `AdvanceStateUpToBlock` from geth.